cilium/CFP-30984-dns-proxy-ha.md

```diff
@@ -0,0 +1,231 @@
+Request :
+```
+message FQDNSelector {
+  string match_name = 1;
+  string match_pattern = 2;
+}
+
+message DNSPolicyRule {
+  string selector_string = 1;
+  repeated FQDNSelector port_rules= 2;
+  repeated string match_labels = 3;
+  repeated uint32 selections = 4;
+}
+
+message DNSPolicyRules {
+  uint64 endpoint_id = 1;
+  uint32 port = 2;
+  repeated DNSPolicyRule rules = 3;
+}
+```
+
+Response :
+```
+message Result  {
+    message string = 1;
+}
+```
+
+### Tracking policy updates to SDP instances
+
+Since SDP can be deployed out of band and users can choose to completely disable built-in proxy to run multiple instances of SDP, agent should be prepared to handle multiple instances. In order to ensure all instances  have upto date policy revisions, agent will maintain a mapping of ACKed policy revision numbers against stream ID.
+Since policy revision numbers are reset when agent restarts, we need to unconditionally send policy updates to SDP on agent restart.
+
+## Impacts / Key Questions
+
+### Getting the DNS Rules
+
+We need the DNS Rules for the Standalone DNS proxy to enforce the L7 DNS policy. The policy are the source of turth for the rules and are propagated to the agent when we apply the policy. We explore the options to get the DNS rules for the DNS proxy.
+
+#### Running the GRPC sever in the agent
+
+We can run a gRPC server in the agent to serve the DNS rules to the DNS proxy. SDP will be responsible for creating the connection with the agent. And once SDP establish a connection agent can keep track of the stream and send the DNS rules to the SDP. Agent can then reuse the same stream to send updates in the DNS rules.
+In case, cilium agent is still not up, SDP will keep trying to connect to the agent until the connection is established.
+
+##### Pros
+
+* SDP instances has the responsibility to connect to the agent.
+* Reusing the same stream will be effiecient in terms of resources as we are not creating/destroying the stream for every update.
+
+##### Cons
+
+* An overhead on the cilium agent to keep track of the streams of the connected SDP instances.
+* Streams are not thread safe, so if we have multiple threads using the same stream we will need to handle the synchronization.
+* If SDP in future decides to handle policy updates as well, it will be tricky. SDP needs to keep trying to sync with CA as well as CA needs to keep a track of the SDP instances to send updates.
+
+![standalone-dns-proxy-up-scenario](./images/standalone-dns-proxy-up-scenario.png)
+![standalone-dns-proxy-down-scenario](./images/standalone-dns-proxy-down-scenario.png)
+
+### Reading from file system on startup
+
+SDP can read from the file system(`ep_config.json`) and get the DNS Rules on bootup. In case, it is able to connect to cilium agent, cilium agent will send the current snapshot of the dns rules to sdp. In case, cilium agent is down, SDP can continue serving the DNS requests based on DNS rules retrieved from the filesystem on bootup.
```

Comment:
> Is this a facility that already exists?

We will need to update the state file with DNS rules as soon as the policy is applied. This PR is trying to address that: https://github.com/cilium/cilium/pull/33412

The SDP always reconcile from the state first and then gets the data from cilium agent through grpc. This helps SDP to have enough information to resolve DNS request even if Cilium agent is not present.
> Any comment on whether this should be part of ep_config.json or stored separately

We can think about storing the data needed by SDP in a separate file but that can be handled during implementation I suppose. 

>  Does this make it more complicated to reconcile with the state from gRPC (or, put more generally, how will SDP track changes over time)?

Assuming Cilium agent will also be reading from the same state. It should not complicate things IMO to reconcile the state from grpc even if we store the rules in a separate file. 

>put more generally, how will SDP track changes over time)?

Not sure what this means. Cilium agent is the source of truth so once SDP is connected to the Cilium agent, the SDP will have the latest rules. In case, agent is not present, SDP still has the local cache to serve the request. Also if SDP restarts, the state file will be used a fallback until agent is back.

Reply:
SDP reading from statefile/maps is only on bootup and then it connects to agent and update its in-mem state. If CA is down, SDP can serve with whatever it read from statefile/maps and this allows existing pods dns requests to be served uninterrupted even when CA goes down

Reply:
My gut feeling here is that all the filesystem stuff is not strictly necessary because having both agents down at the same time already broke HA. It would reduce the scope of the CFP and move it forward to just defer all the filesystem stuff to a subsequent milestone and re-evaluate how hard the requirement is for that aspect of the implementation. Just thinking from a simplicity perspective and how the implementation achieves the core goals. I can see some potential for more resilience with this, I just think that it's probably overthinking the problem at this stage if the basics are not yet upstream.